### DISCUSSION TITLE
Passing down data from parent model to child model

### DISCUSSION THREAD

**Discussion by HarishHary** [pydantic/pydantic#6587] [Question]
Hello,

I would like to know what is the best way of passing down data from parent model to child model? or across model for validation purposes.

Let's say I have the following models:

```
class Default(BaseModel, extra='forbid', validate_assignment=True):
    image_references: ImageReferences

class Stacks(BaseModel, extra='forbid', validate_assignment=True):
    linux: LinuxStack = {}

class LinuxStack(BaseModel, extra='forbid', validate_assignment=True):
    default: LinuxDefaultHostData
    hosts: list[LinuxHost]

class LinuxHost(Host):
    user: LinuxDefaultUser
    tags: Tags = Tags(type=TypeEnum.linux)

class Host(BaseModel, extra='forbid', validate_assignment=True):
    name: str
    settings: dict = {}
    image_reference: str | ImageReference
```

How do I check from the `Host` model that the ImageReference exists? ImageRefence is from a external model called `Default` that is loaded by reading a yaml file.

`Stacks` is loaded as well reading another yaml file. `Default` as the name suggest hold all the default data that can be used to check the data from the `Stacks` model.

Not totally sure on what's the best way to proceed here. Validation context is somehow close to what I am trying to achieve here.

https://docs.pydantic.dev/dev-v2/usage/validators/#validation-context

But I am not totally sure on how to use it.

This is how I do it so far:
```
class LinuxStack(WithAllowedData):
    default: LinuxDefaultHostData
    hosts: list[LinuxHost]

    def __init__(self, **data):
        default = data.get("default", {})
        hosts = data.get("hosts", [])
        updated_hosts = [deep_update(default, host) for host in hosts]
        data.update({"hosts": updated_hosts})
        with init_context({**_init_context_var.get(), **{"type": TypeEnum.linux}}):
            super().__init__(**data)
```

Not sure if this is really nice to override `__init__` function...

Any OOP pattern or any way to handle this case in a proper way?

**Comment by estheruary** [comment]
Just ran into this, here's the solution I came up with.

```python
from contextvars import ContextVar
from typing import Any

_parent_data = ContextVar("_parent_data")

class MyBase(BaseModel):
  def __init__(self, /, **data: Any):
    super().__init__(**data)
    self._parent_data = _parent_data.get(None)

  @classmethod
  @contextmanager
  def bind(cls, parent_data):
    token = _parent_data.set(parent_data)
    try:
       yield cls
    finally:
        _parent_data.reset(token)
```

Then you use it like this.

```python
class Parent(MyBase):
  a: int
  b: "Child"

class Child(MyBase):
  c: int

with Parent.bind(the_data_to_pass):
  Parent.model_validate(obj)
```

Now no matter how deep the nesting child models can get `self._parent_data`.

**Reply by NicholasPini** [reply]
I tried this method, but I have a problem: in validators, I cannot access the context passed to model_validate. The ValidationInfo object has the `context` variable set to None. If I remove the overload of the __init__ function, the context is there.
